File: backend/app/schemas/receipt.py
"""Receipt schemas."""

from __future__ import annotations

import datetime as dt
from decimal import Decimal
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel


class ReceiptStatus(StrEnum):
    """Processing lifecycle of a receipt (mirrors the DB CHECK constraint)."""

    PENDING = "pending"
    PROCESSING = "processing"
    DONE = "done"
    FAILED = "failed"


class ReceiptImage(BaseModel):
    """One page of a receipt (``public.receipt_images``).

    ``image_url`` is the Storage object key, not a fetchable URL — short-lived signed URLs
    are minted on read (design decision 19).
    """

    id: UUID
    receipt_id: UUID
    image_url: str
    page_number: int
    created_at: dt.datetime


class ReceiptItem(BaseModel):
    """A parsed line item (``public.receipt_items``) — the bill-detail read model.

    ``category`` is the resolved category *name* (e.g. "Dairy"), not the id, so the client
    can render it without a second lookup. ``needs_review`` flags a low-confidence line for
    the UI to highlight (MyBill.md §6).
    """

    id: UUID
    name: str
    brand: str | None = None
    category: str | None = None
    quantity: Decimal
    unit: str | None = None
    unit_price: Decimal
    total_price: Decimal
    ocr_confidence: float | None = None
    needs_review: bool = False


class ItemSearchResult(BaseModel):
    """A line item matched by search, with the bill it belongs to (MyBill.md §5).

    Carries enough receipt context (store, date, ``receipt_id``) for the client to show a
    useful result row and link back to the full bill.
    """

    id: UUID
    receipt_id: UUID
    name: str
    category: str | None = None
    quantity: Decimal
    unit: str | None = None
    unit_price: Decimal
    total_price: Decimal
    store_name: str | None = None
    date: dt.date | None = None


class Receipt(BaseModel):
    """A receipt with its pages and (once parsed) its summary fields.

    A receipt holds 1..N images (decision 24): the first is created at upload, further
    pages are appended for receipts too long to photograph in one shot. The parsed summary
    fields (store, date, totals) are null until OCR completes; line items are read
    separately via the items endpoint rather than joined onto every list row.
    """

    id: UUID
    status: ReceiptStatus
    created_at: dt.datetime
    # Parsed summary — populated from the row once status is done; null while pending.
    store_name: str | None = None
    date: dt.date | None = None
    time: dt.time | None = None
    total: Decimal | None = None
    tax: Decimal | None = None
    discount: Decimal | None = None
    payment_method: str | None = None
    ocr_confidence: float | None = None
    # Number of parsed line items, for the list view. None when not populated.
    item_count: int | None = None
    # Sum of the line items — an estimate of the bill when OCR found no printed total
    # (e.g. the total was cut off in the photo). Distinct from ``total`` (the printed
    # amount) so the client can label an estimate as such.
    item_total: Decimal | None = None
    # Ordered by page number. A receipt created via upload always has at least one.
    images: list[ReceiptImage] = []

    @property
    def page_count(self) -> int:
        return len(self.images)

File: mobile/lib/features/scan/domain/receipt.dart
/// Processing lifecycle of a receipt. Mirrors the backend's `ReceiptStatus` and the
/// database CHECK constraint.
enum ReceiptStatus {
  pending,
  processing,
  done,
  failed;

  /// Unknown values map to [pending] rather than throwing: the backend may add a state
  /// (Phase 2 OCR) before the app ships an update, and a new status is not a reason to
  /// fail an upload the server already accepted.
  static ReceiptStatus parse(String? value) => ReceiptStatus.values.firstWhere(
    (status) => status.name == value,
    orElse: () => ReceiptStatus.pending,
  );
}

/// One page of a receipt.
class ReceiptImage {
  const ReceiptImage({
    required this.id,
    required this.imageUrl,
    required this.pageNumber,
  });

  factory ReceiptImage.fromJson(Map<String, dynamic> json) => ReceiptImage(
    id: json['id'] as String,
    // The storage object key, not a fetchable URL — signed URLs are minted on read
    // in Phase 3 (design decision 19).
    imageUrl: json['image_url'] as String,
    pageNumber: json['page_number'] as int? ?? 1,
  );

  final String id;
  final String imageUrl;
  final int pageNumber;
}

/// A parsed line item on a receipt (`GET /receipts/{id}/items`).
class ReceiptItem {
  const ReceiptItem({
    required this.id,
    required this.name,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
    this.brand,
    this.category,
    this.unit,
    this.needsReview = false,
  });

  factory ReceiptItem.fromJson(Map<String, dynamic> json) => ReceiptItem(
    id: json['id'] as String,
    name: json['name'] as String? ?? '',
    brand: json['brand'] as String?,
    category: json['category'] as String?,
    quantity: _toDouble(json['quantity']) ?? 1,
    unit: json['unit'] as String?,
    unitPrice: _toDouble(json['unit_price']) ?? 0,
    totalPrice: _toDouble(json['total_price']) ?? 0,
    needsReview: json['needs_review'] as bool? ?? false,
  );

  final String id;
  final String name;
  final String? brand;
  final String? category;
  final double quantity;
  final String? unit;
  final double unitPrice;
  final double totalPrice;

  /// True for a low-confidence parse the UI should highlight for correction.
  final bool needsReview;
}

/// A receipt: its pages and, once OCR completes, its parsed summary (task 1.3.5,
/// decision 24; Phase 3 bill viewer).
///
/// A receipt holds 1..N pages — a long receipt is photographed in several shots and the
/// pages are appended to the same bill. The parsed summary fields are null until the
/// pipeline settles the receipt on `done`; line items are fetched separately.
class Receipt {
  const Receipt({
    required this.id,
    required this.status,
    required this.createdAt,
    this.storeName,
    this.date,
    this.time,
    this.total,
    this.tax,
    this.discount,
    this.paymentMethod,
    this.images = const [],
  });

  factory Receipt.fromJson(Map<String, dynamic> json) => Receipt(
    id: json['id'] as String,
    status: ReceiptStatus.parse(json['status'] as String?),
    createdAt:
        DateTime.tryParse(json['created_at'] as String? ?? '') ??
        DateTime.now(),
    storeName: json['store_name'] as String?,
    date: DateTime.tryParse(json['date'] as String? ?? ''),
    // Backend sends "HH:MM:SS"; keep just HH:MM for display.
    time: _shortTime(json['time'] as String?),
    total: _toDouble(json['total']),
    tax: _toDouble(json['tax']),
    discount: _toDouble(json['discount']),
    paymentMethod: json['payment_method'] as String?,
    images: ((json['images'] as List<dynamic>?) ?? const [])
        .map((i) => ReceiptImage.fromJson(i as Map<String, dynamic>))
        .toList(),
  );

  final String id;
  final ReceiptStatus status;
  final DateTime createdAt;

  /// Parsed summary — null until OCR completes.
  final String? storeName;
  final DateTime? date;
  final String? time;
  final double? total;
  final double? tax;
  final double? discount;
  final String? paymentMethod;

  final List<ReceiptImage> images;

  int get pageCount => images.length;
}

/// Parse a numeric field that the backend may serialise as either a JSON number or a
/// string (pydantic renders Decimal as a string).
double? _toDouble(Object? value) => switch (value) {
  null => null,
  final num n => n.toDouble(),
  final String s => double.tryParse(s),
  _ => null,
};

/// "18:42:00" → "18:42"; null/blank stays null.
String? _shortTime(String? value) {
  if (value == null || value.isEmpty) return null;
  final parts = value.split(':');
  return parts.length >= 2 ? '${parts[0]}:${parts[1]}' : value;
}
